server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1524,6 +1524,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
 
     private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean checkRefreshPending) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
+        if (request.source() != null && (request.source().query() instanceof MatchNoneQueryBuilder) == false) {
+            final SearchSourceBuilder rewritten = Rewriteable.rewrite(
+                request.source(),
+                indicesService.getMappingAwareRewriteContextProvider(
+                    request.shardId().getIndex(),
+                    request::nowInMillis,
+                    request.shardId(),
+                    request.shardRequestIndex(),
+                    request.getClusterAlias(),
+                    request.getRuntimeMappings().isEmpty() == false
+                ).get()
+            );
+            if (rewritten.query() instanceof MatchNoneQueryBuilder) {
+                return new CanMatchShardResponse(false, null);
+            }
+        }
```

Comment:
The `MappingAwareRewriteContext` doesn't seem to be needed. Why not creating a new `SearchExecutionContext` when the index service is [available](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1567).
You should also reuse `queryStillMatchesAfterRewrite(request, context)` to ensure that all cases are checked (alias filter and global aggregations).

Reply:
> If we use SearchExecutionContext then also range queries (and maybe others?) could be resolved to MatchNoneQueryBuilder instance.

The initial change already used `SearchExecutionContext` but it sets the searcher to `null`. That's why I said that the special context is not needed. There's some logic to handle `SearchExecutionContext` with a null searcher, we should add some javadocs to explain the use case (rewriting with the mapping only) but it works.

Reply:
Right, I see this now. Yes, if we pass a null searcher then this should work. 👍

Reply:
I had to add an additional null check in `isFieldWithinQuery`. That logic uses the reader which might be null in this case and triggers a NPE at some point.